contracts/RegistryV1.sol

```diff
@@ -0,0 +1,313 @@
+   * @param entity The entity address
+   * @return bool True if the entity requires approval, false otherwise
+   */
+  function getRequiresApprovalForRewardsAgreements(
+    address entity
+  ) external view returns (bool) {
+    return _requiresApproval[_entityToId[entity]];
+  }
+
+  /**
+   * @notice Set whether a Rewards Entity requires approval for agreements
+   */
+  function setRequiresApprovalForRewardsAgreements(
+    bool requiresApproval
+  ) external {
+    uint256 entityId = _entityToId[msg.sender];
+    if (entityId == 0) {
+      revert EntityDoesNotExist(msg.sender);
+    }
+
+    _requiresApproval[entityId] = requiresApproval;
+    emit RequiresApprovalForRewardsAgreements(msg.sender, requiresApproval);
+  }
+
+  /**
+   * @notice Registers a Rewards Consumer - Rewards Provider relationship between two Rewards Entities, should be called by the Rewards Consumer
+   * @param rewardsProvider The provider entity address
+   */
+  function registerRewardsAgreement(address rewardsProvider) external {
+    uint256 providerId = _entityToId[rewardsProvider];
+    uint256 consumerId = _entityToId[msg.sender];
+
+    if (providerId == 0 || consumerId == 0) {
+      revert EntityDoesNotExist(providerId == 0 ? rewardsProvider : msg.sender);
+    }
+
+    // If the provider requires approval, revert the transaction
+    if (_requiresApproval[providerId]) {
+      revert ProviderRequiresApproval(rewardsProvider);
+    }
+
+    // Check if agreement already exists
+    bytes32 agreementKey = keccak256(abi.encodePacked(providerId, consumerId));
+    if (_agreements[agreementKey]) {
+      revert AgreementAlreadyExists(rewardsProvider, msg.sender);
+    }
+
+    _agreements[agreementKey] = true;
+    emit RewardsAgreementRegistered(rewardsProvider, msg.sender);
+  }
+
+  /**
+   * @notice Approve a rewards agreement, should be called by the Rewards Provider
+   * @param rewardsConsumer The consumer entity address
+   */
+  function approveRewardsAgreement(address rewardsConsumer) external {
+    uint256 providerId = _entityToId[msg.sender];
+    uint256 consumerId = _entityToId[rewardsConsumer];
+
+    if (providerId == 0 || consumerId == 0) {
+      revert EntityDoesNotExist(providerId == 0 ? msg.sender : rewardsConsumer);
+    }
+
+    // Create the agreement
+    bytes32 agreementKey = keccak256(abi.encodePacked(providerId, consumerId));
+    _agreements[agreementKey] = true;
+    emit RewardsAgreementApproved(msg.sender, rewardsConsumer);
+  }
+
+  /**
+   * @notice Registers a user as being referred to a rewards agreement
+   * @param user The address of the user being referred
+   * @param userSignature The user's signature authorizing the referral
+   * @param rewardsConsumer The address of the rewards consumer entity
+   * @param rewardsProvider The address of the rewards provider entity
+   * @return success Whether the referral was successfully registered
+   */
+  function registerReferral(
```

Comment:
I'm evolving my thoughts about the need for batch processing referrals.
If we don't do batch processing, we could go back to normal revert reasons and make this function's behavior more standard/predictable.

I'll post more on this on Slack relayed to the research on relayers.